balanced-optiml/bal-optiml-library/library.whizzml

```diff
@@ -0,0 +1,139 @@
+              "code" 102}))))
+
+;; Function that splits a dataset in training, validation and test datasets
+(define (_create-train-validation-test ds-id train-rate seed)
+  (let (first-split (create-dataset-split ds-id train-rate seed)
+        second-split (create-dataset-split (last first-split) 0.5 seed))
+    (cons (head first-split) second-split)))
+
+
+;; Function to apply stratified split for the rare class of the objective field
+;; with a given train
+(define (rare-stratified-split-fn ds-id training-rate objective-field
+                                  rare-class project)
+  (let (dataset (fetch ds-id)
+        ds-name (dataset "name")
+        fields (resource-fields dataset)
+        _ (_check-cat fields objective-field rare-class)
+        objective-id ((find-field fields objective-field) "id")
+        filters [(flatline "(= (f {{objective-id}}) {{rare-class}})")
+                 (flatline "(!= (f {{objective-id}}) {{rare-class}})")]
+        names [rare-class (str "not " rare-class)]
+        cat-ds-ids (iterate (acc [] expr filters name names)
+                     (append acc
+                       (_create-train-validation-test
+                         (create-dataset {"origin_dataset" ds-id
+                                          "lisp_filter" expr
+                                          "name" (str (dataset "name")
+                                                      " - "
+                                                      name)
+                                          "project" project})
+                         training-rate
+                         "bigml")))
+        inner-ds (flatten cat-ds-ids)
+        ds-ids (map (lambda (ids name)
+                      (merge-datasets ids {"name" name}))
+                    (_transpose cat-ds-ids)
+                    [(str ds-name " - train")
+                     (str ds-name " - validation")
+                     (str ds-name " - test")]))
+    (wait* ds-ids)
+    (map _safe-delete inner-ds)
+    ds-ids))
+
+
+;; Function to generated datasets that keep all instances of the rare class
+;; while undersampling the instances in the common class to maintain a
+;; proportion with the rare class
+(define (undersample-fn ds-id number-of-samples objective-field rare-class
+         proportion)
+  (let (dataset (fetch ds-id)
+        ds-name (dataset "name")
+        fields (resource-fields dataset)
+        _ (_check-cat fields objective-field rare-class)
+        objective-id ((find-field fields objective-field) "id")
+        rows (dataset "rows")
+        field-summary (field-distribution (find-field fields objective-field))
+        _ (log-info field-summary)
+        rare-instances ((head (filter (lambda (f) (= (f 0) rare-class))
+                                      field-summary)) 1)
+        common-instances (- rows rare-instances)
+        balance-rate (* proportion (/ rare-instances common-instances))
```

Comment:
is there any possibility (perhaps by a badly specified rare-class) that this gives a division by zero?

Reply:
The only corner case is iff there's only instances that belong to the rare class in the dataset.  I'll add a check to avoid that case and another to prevent using datasets with less that 10 instances in the rare class to ensure that enough instances are provided in the test and validation datasets under reasonable conditions.